{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This script is used for register inference. Can also visualize clusters."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import random\n",
    "from scipy.special import softmax\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "from PIL import Image"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Take a list of (sequences, counts) and convert them into one-hot encodings\n",
    "def encodeSequences(seqs):\n",
    "    dic = {}\n",
    "    v0 = np.zeros(20)\n",
    "    for c, v in zip(\"ACDEFGHIKLMNPQRSTVWY\", np.eye(20)):\n",
    "        dic[c] = v\n",
    "    encSeqs = []\n",
    "    for seq, ct in seqs:\n",
    "        # Pad entries to allow shifting to work - padding at C-terminal end \n",
    "        encSeqs.append(( np.array([dic[c] for c in seq] + [v0]*(n_padded-(15-n_clusters+1))), ct )) \n",
    "\n",
    "    return encSeqs\n",
    "\n",
    "# Updates all p1 positions\n",
    "def iteration(model, seqs, beta, background):\n",
    "    p1selection = list(range(0,n_clusters)) #No trash cluster\n",
    "    for seqinfo in seqs:\n",
    "        seq, n, shift = seqinfo\n",
    "        if shift != -1:\n",
    "            model -= seq[shift:shift+n_padded] * n\n",
    "        sums = np.sum(model, axis = 1).reshape((-1,1))\n",
    "        pwm = np.log(model/sums) - background\n",
    "        p = [] #[0]  #No trash cluster\n",
    "        \n",
    "        for i in range(n_clusters):\n",
    "            p.append( n * beta * np.sum( pwm * seq[i:i+n_padded] ) )\n",
    "        shift = random.choices( p1selection, weights = softmax(p), k = 1 )[0]\n",
    "        seqinfo[2] = shift\n",
    "        if shift != -1:\n",
    "            model += seq[shift:shift+n_padded] * n\n",
    "\n",
    "# Updates all p1 positions with beta = inf (0 temperature)\n",
    "def finalIteration(model, seqs, background):\n",
    "    p1selection = list(range(0,n_clusters))\n",
    "    for seqinfo in seqs:\n",
    "        seq, n, shift = seqinfo\n",
    "        if shift != -1:\n",
    "            model -= seq[shift:shift+n_padded] * n\n",
    "        sums = np.sum(model, axis = 1).reshape((-1,1))\n",
    "        pwm = np.log(model/sums) - background\n",
    "        p = [] #make empty to start this time\n",
    "        for i in range(n_clusters):\n",
    "            p.append( n * np.sum( pwm * seq[i :i+n_padded ] ) ) \n",
    "            \n",
    "        shift = p1selection[np.argmax(p)]\n",
    "        seqinfo[2] = shift\n",
    "        if shift != -1:\n",
    "            model += seq[shift:shift+n_padded] * n\n",
    "        \n",
    "            \n",
    "def reportScore(model, seqs, background):\n",
    "    sums = np.sum(model, axis = 1).reshape((-1,1))\n",
    "    pwm = np.log(model/sums) - background\n",
    "    \n",
    "    score = 0\n",
    "    for seq, n, shift in seqs:\n",
    "        if shift != -1:\n",
    "            score += np.sum( pwm * seq[shift:shift+n_padded] ) * n\n",
    "            \n",
    "    pd.DataFrame(pwm).to_csv('{}_{}_pwm.csv'.format(save_prefix, allele)) #overwrites until the final time and save pwm\n",
    "    pd.DataFrame(model).to_csv('{}_{}_model.csv'.format(save_prefix, allele))\n",
    "    return score\n",
    "    \n",
    "# seqs : list of (sequence, count) tuples.\n",
    "# iterations : number of iterations to carry out (not including the final iteration)\n",
    "# betamin, betamax : the first iteration will be done with beta=betamin, and one before the final iteration\n",
    "# with beta=betamax. Beta is linearly interpolated linearly in between\n",
    "# Beta is the inverse of temprature, so this simulates cooling.\n",
    "\n",
    "\n",
    "def alignSequences(seqs, iterations, betamin, betamax):\n",
    "\n",
    "    model = np.ones((n_padded,20))\n",
    "    result = []\n",
    "    assignments = np.random.randint(0,n_clusters,len(seqs)) #GOT RID OF TRASH CLUSTER (no -1)\n",
    "    for (seq, n), shift in zip(encodeSequences(seqs), assignments):\n",
    "        if shift != -1:\n",
    "            model += seq[shift:shift+n_padded] * n\n",
    "        result.append([seq, n, shift])\n",
    "    background = np.log(1/20)\n",
    "    \n",
    "    betas = np.linspace(betamin,betamax,iterations)\n",
    "    for i,beta in enumerate(betas):\n",
    "        iteration(model, result, beta, background)        \n",
    "        print (\"Iteration {}, beta = {}, score = {}\".format(i, beta, reportScore(model, result, background)))\n",
    "    \n",
    "    finalIteration(model, result, background)\n",
    "    print (\"Final iteration, score = {}\".format(reportScore(model, result, background)))\n",
    "    \n",
    "    return result\n",
    "\n",
    "def readInput(filename):\n",
    "    seqs = []\n",
    "    with open(filename, 'rt') as fin:\n",
    "        for i, line in enumerate(fin):\n",
    "            if i == 0: continue\n",
    "            seq, ct = line.split(',')\n",
    "            seqs.append((seq, int(ct)))\n",
    "    return seqs\n",
    "\n",
    "def readInput_alt(df):\n",
    "    #Alternative: using dataframes and ignoring peptide counts\n",
    "    #ignoring counts - i.e. count=1\n",
    "    seqs = []\n",
    "    for entry in df['name']:\n",
    "        seq, ct = entry, 1\n",
    "        seqs.append((seq, int(ct)))\n",
    "    return seqs\n",
    "\n",
    "\n",
    "def writeOutput(filename, alignment, sequences):\n",
    "    lines = [\"sequence,count,P1\"]\n",
    "    for (_,ct,p1), (seq,_) in zip(alignment, sequences):\n",
    "        lines.append(\"{},{},{}\".format(seq,ct,p1))\n",
    "    with open(filename, 'wt') as fout:\n",
    "        fout.write('\\n'.join(lines))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#functions to make sequence logos\n",
    "\n",
    "global_aa = list(\"ACDEFGHIKLMNPQRSTVWY\")\n",
    "polars = set(['G', 'S', 'T', 'Y', 'C', 'Q', 'N'])\n",
    "basics = set(['K', 'R', 'H'])\n",
    "acids = set(['D', 'E'])\n",
    "#The rest are hydrophobic and black\n",
    "\n",
    "def readimg(fname, col):\n",
    "    arr = plt.imread(fname)\n",
    "    newarr = np.zeros(shape=(len(arr), len(arr[0]), 4), dtype = \"uint8\")\n",
    "    uparr = np.zeros(shape=(len(arr), len(arr[0]), 4), dtype = \"uint8\")\n",
    "    downarr = np.zeros(shape=(len(arr), len(arr[0]), 4), dtype = \"uint8\")\n",
    "    for x in range(len(arr)):\n",
    "        for y in range(len(arr[0])):\n",
    "            newarr[x][y][3] = 255 - arr[x][y][0] * 255\n",
    "            newarr[x][y][0] = col[0]\n",
    "            newarr[x][y][1] = col[1]\n",
    "            newarr[x][y][2] = col[2]\n",
    "    return newarr\n",
    "\n",
    "imdic = {}\n",
    "updic = {}\n",
    "downdic = {}\n",
    "for i in range(26):\n",
    "    c = chr(i + ord('A'))\n",
    "    col = (0,0,0)\n",
    "    if c in polars:\n",
    "        col = (0,192,0)\n",
    "    elif c in basics:\n",
    "        col = (24,32,255)\n",
    "    elif c in acids:\n",
    "        col = (255,0,0)\n",
    "        \n",
    "    imdic[c] = readimg(\"{}_crop.png\".format(c), col)\n",
    "    #directory with images of letters in it\n",
    "    \n",
    "def drawColumn(ax, seqs, hgts, shift, wide):\n",
    "    i = 0\n",
    "    for i,x in enumerate(range(1,16)):\n",
    "        l = x-wide/2 + shift\n",
    "        r = x+wide/2 + shift\n",
    "        b = 0\n",
    "        for c,p in seqs[i][::-1]:\n",
    "            hgt = hgts[i] * p\n",
    "            t = b + hgt\n",
    "            if hgt == 0:\n",
    "                b = t\n",
    "                continue\n",
    "                \n",
    "            mgn = hgt * 0\n",
    "            ax.imshow(imdic[c], aspect = \"auto\", extent = (l,r,b+mgn,t-mgn), interpolation = 'bilinear')\n",
    "            b = t\n",
    "\n",
    "    ax.set_xticks(range(1,16))\n",
    "    ax.set(xlim = (0,16), ylim = (0,max(hgts)*1.1))\n",
    "\n",
    "    \n",
    "def drawLogo(seqlist):\n",
    "    fig, ax = plt.subplots(1,1,figsize=(10, 6))\n",
    "    w = 0.8#0.25\n",
    "    ws = 0.28\n",
    "    h = 0\n",
    "    for i in range(-1,0):\n",
    "        fractions, heights = seqlist[i+1]\n",
    "        drawColumn(ax, fractions, heights, ws*(i+1), w)\n",
    "        h = max(h, max(heights))\n",
    "    ax.set(ylim = (0, 1.1*h), xlim = (0.5, 15.5))\n",
    "    \n",
    "    xtc = []\n",
    "    xlab = []\n",
    "    for i in range(1,16):\n",
    "        for j in range(0,1):\n",
    "            pos = i + j*ws\n",
    "            lab = \"P{}\".format(i)\n",
    "            xtc.append(pos)\n",
    "            xlab.append(lab)\n",
    "    ax.set_xticks(xtc)\n",
    "    ax.set_xticklabels(xlab, rotation=270)\n",
    "    ax.set_ylabel(\"Bits\")\n",
    "    \n",
    "    ax.spines['right'].set_visible(False)\n",
    "    ax.spines['top'].set_visible(False)\n",
    "    \n",
    "    plt.title(str(cluster))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Process data\n",
    "save_prefix = 'test-v4'\n",
    "\n",
    "n_padded = 9 #size of kernel (dimension in model)\n",
    "n_clusters = 11 #number of different clusters that can assign to\n",
    "\n",
    "##\n",
    "data = pd.read_csv('../twist_covid_filtered+normalized_counts.csv').fillna(0)\n",
    "data.drop(columns='Unnamed: 0',inplace=True)\n",
    "data.drop(columns='Unnamed: 0.1',inplace=True)\n",
    "data=data[data['doped_match']!='match'] #keep only covid peptides\n",
    "\n",
    "##\n",
    "for allele in ['401','402','404']:\n",
    "\n",
    "    doped_positive = (data['count_ doped-{} post-R3 '.format(allele)]>=10)\n",
    "    undoped_positive = (data['frac_ undoped-{} post-R2 - positive '.format(allele)]>data['frac_ undoped-{} post-R1 - positive '.format(allele)])\n",
    "\n",
    "    data['hit d+u'] = doped_positive & undoped_positive\n",
    "\n",
    "    sequences = readInput_alt(data[data['hit d+u']==True])\n",
    "    alignment = alignSequences(sequences, 60, 0.05, 1)\n",
    "    fname = \"{}_{}_twist_covid_filtered+normalized_counts_{}.csv\".format(save_prefix,allele,'d+u')\n",
    "    writeOutput(fname, alignment, sequences)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "## Visualize\n",
    "\n",
    "# Each sequence is a tuple of (fraction data, height data)\n",
    "# Fraction data is a list of 9 tuples of (amino acid, relative height)\n",
    "# Height data is a list of 9 numbers denoting height of the column \n",
    "\n",
    "#re-imports data\n",
    "for allele in ['401','402','404']:\n",
    "    \n",
    "    fname = \"{}_{}_twist_covid_filtered+normalized_counts_{}.csv\".format(save_prefix,allele,'d+u')\n",
    "    data_clustered = pd.read_csv(fname)\n",
    "\n",
    "    for cluster in range(0,n_clusters):\n",
    "        data_sub = data_clustered[data_clustered['P1']==cluster]\n",
    "\n",
    "        frac_list = []    \n",
    "        height_list = []\n",
    "\n",
    "        for j in range(15): #iterate over positions\n",
    "            temp = []\n",
    "            height = np.log2(20)\n",
    "            for c in global_aa:\n",
    "                frac = np.sum(data_sub['sequence'].str[j]==c)/len(data_sub)\n",
    "                temp.append((c,frac))\n",
    "                if frac != 0:\n",
    "                    height += frac*np.log2(frac)\n",
    "            height_list.append(height)\n",
    "            frac_list.append(temp) \n",
    "\n",
    "        drawLogo([(frac_list, height_list)])\n",
    "\n",
    "    #Count the number of sequences in each register\n",
    "    for p1 in list(range(0,n_clusters)):\n",
    "        n = np.sum(data_clustered['P1']==p1)\n",
    "        frac = n/len(data_clustered)\n",
    "        print(\"Register {}: {} sequences; {}%\".format(p1,n,round(frac*100,2))) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
