{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This script is for processing the fasta files from PandaSeq and generating peptide lists. \n",
    "Filters for peptides which are in the ordered library or matched the pattern of the doped sequences."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import csv\n",
    "import string\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import filecmp\n",
    "from scipy import stats\n",
    "import re\n",
    "import fnmatch\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#import the list of DNA sequences that we ordered from Twist\n",
    "merged = pd.read_csv('../nCov_proteome_15mers+SARS-CoV-1 N+S_encoded FINAL remove flanking DNA.csv') \n",
    "merged = merged.rename(columns={'aa':'name'})\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def translate(pep_dna):\n",
    "#Translate peptide to amino acids\n",
    "    codons_aa = []\n",
    "    codons_dna = []\n",
    "    for j in range(0,len(pep_dna),3):   \n",
    "        codons_dna.append(pep_dna[j:j+3])\n",
    "    codons_aa = [codon_2_aa[x] for x in codons_dna] # translate\n",
    "    pep_aa=''.join(codons_aa) #puts letters into string\n",
    "    return pep_aa\n",
    "\n",
    "codon_2_aa = {\"TTT\":\"F\", \"TTC\":\"F\", \"TTA\":\"L\", \"TTG\":\"L\",\n",
    "    \"TCT\":\"S\", \"TCC\":\"S\", \"TCA\":\"S\", \"TCG\":\"S\",\n",
    "    \"TAT\":\"Y\", \"TAC\":\"Y\", \"TAA\":\"*\", \"TAG\":\"*\",\n",
    "    \"TGT\":\"C\", \"TGC\":\"C\", \"TGG\":\"W\", \"TGA\":\"*\",\n",
    "    \"CTT\":\"L\", \"CTC\":\"L\", \"CTA\":\"L\", \"CTG\":\"L\",\n",
    "    \"CCT\":\"P\", \"CCC\":\"P\", \"CCA\":\"P\", \"CCG\":\"P\",\n",
    "    \"CAT\":\"H\", \"CAC\":\"H\", \"CAA\":\"Q\", \"CAG\":\"Q\",\n",
    "    \"CGT\":\"R\", \"CGC\":\"R\", \"CGA\":\"R\", \"CGG\":\"R\",\n",
    "    \"ATT\":\"I\", \"ATC\":\"I\", \"ATA\":\"I\", \"ATG\":\"M\",\n",
    "    \"ACT\":\"T\", \"ACC\":\"T\", \"ACA\":\"T\", \"ACG\":\"T\",\n",
    "    \"AAT\":\"N\", \"AAC\":\"N\", \"AAA\":\"K\", \"AAG\":\"K\",\n",
    "    \"AGT\":\"S\", \"AGC\":\"S\", \"AGA\":\"R\", \"AGG\":\"R\",\n",
    "    \"GTT\":\"V\", \"GTC\":\"V\", \"GTA\":\"V\", \"GTG\":\"V\",\n",
    "    \"GCT\":\"A\", \"GCC\":\"A\", \"GCA\":\"A\", \"GCG\":\"A\",\n",
    "    \"GAT\":\"D\", \"GAC\":\"D\", \"GAA\":\"E\", \"GAG\":\"E\",\n",
    "    \"GGT\":\"G\", \"GGC\":\"G\", \"GGA\":\"G\", \"GGG\":\"G\"}\n",
    "\n",
    "#round/allele correspondence of files\n",
    "plex_dict = {\n",
    "    '7096':' doped-401 R0 (doped)',\n",
    "    '7097':' doped-402 R0 (doped)',\n",
    "    '7098':' doped-404 R0 (doped)',\n",
    "    '7099':' doped-401 post-R1 ',\n",
    "    '7100':' doped-402 post-R1 ',\n",
    "    '7101':' doped-404 post-R1 ',\n",
    "    '7102':' doped-401 post-R2 ',\n",
    "    '7103':' doped-402 post-R2 ',\n",
    "    '7104':' doped-404 post-R2 ',\n",
    "    '7105':' doped-401 post-R3 ',\n",
    "    '7106':' doped-402 post-R3 ',\n",
    "    '7107':' doped-404 post-R3 ',\n",
    "    '7108':' doped-401 post-R4 ',\n",
    "    '7109':' doped-402 post-R4 ',\n",
    "    '7110':' doped-404 post-R4 ',\n",
    "    '7111':' undoped-401 R0 (undoped)',\n",
    "    '7112':' undoped-402 R0 (undoped)',\n",
    "    '7113':' undoped-404 R0 (undoped)',\n",
    "    '7114':' undoped-401 post-R1 - positive ',\n",
    "    '7115':' undoped-402 post-R1 - positive ',\n",
    "    '7116':' undoped-404 post-R1 - positive ',\n",
    "    '7117':' undoped-401 post-R1 - negative ',\n",
    "    '7118':' undoped-402 post-R1 - negative ',\n",
    "    '7119':' undoped-404 post-R1 - negative ',\n",
    "    '7120':' undoped-401 post-R2 - positive ',\n",
    "    '7121':' undoped-402 post-R2 - positive ',\n",
    "    '7122':' undoped-404 post-R2 - positive ',\n",
    "    '7123':' undoped-401 post-R2 - negative ',\n",
    "    '7124':' undoped-402 post-R2 - negative ',\n",
    "    '7125':' undoped-404 post-R2 - negative ',\n",
    "    '7126':' undoped-401 post-R3 - positive ',\n",
    "    '7127':' undoped-402 post-R3 - positive ',\n",
    "    '7128':' undoped-404 post-R3 - positive ',\n",
    "    '7129':' undoped-401 post-R3 - negative ',\n",
    "    '7130':' undoped-402 post-R3 - negative ',\n",
    "    '7131':' undoped-404 post-R3 - negative '\n",
    "    }\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "flank5_seq = 'TATTGCTAGCGTTTTGGCAGCT'\n",
    "flank3_seq_401 = 'GGTGGATCCGGTGGCGGAGAACAAAAATTAATTAGTGAAGAAGATTTAGGCGGTCTAGAAGTTCTGTTCCAGGGGCCCGGTGGCGGGTCCGGCGGT'\n",
    "flank3_seq_402 = 'GGTGGATCCGGTGGCGGAGAACAAAAATTAATTAGTGAGGAGGACCTTGGCGGTCTAGAAGTTCTGTTCCAGGGGCCCGGTGGCGGGTCCGGCGGT'\n",
    "flank3_seq_404 = 'GGTGGATCCGGTGGCGGAGAACAAAAATTAATCTCAGAGGAAGATCTTGGCGGTCTAGAAGTTCTGTTCCAGGGGCCCGGTGGCGGGTCCGGCGGT'\n",
    "flank_stop_splicevar = 'TATTGCTAGCGTTTTGGCAGCTGGATAAGCTGGTGTTTAGCGCTGGTTGCTGTGAGTGCCCGGTGGATCCGGTGGCGGGTCCGGCGGT'#variant seen in Sanger sequence of library for doping into;orig peptide+linker spliced out\n",
    "\n",
    "def process(plex):\n",
    "    file_name = 'twist_covid_'+plex+'_ps.fasta'\n",
    "    dna = []\n",
    "    aa = []\n",
    "\n",
    "    number_contigs = 0\n",
    "\n",
    "    with open(file_name, 'r') as read_file:  \n",
    "        for line in read_file:\n",
    "            if line[0] != '>':\n",
    "                number_contigs +=1\n",
    "                line = line.strip()\n",
    "                \n",
    "                #check if sequence contains the exact flanking region that should and remove:\n",
    "                flank5 = line.find(flank5_seq)\n",
    "                flank3_401 = line.find(flank3_seq_401) \n",
    "                flank3_402 = line.find(flank3_seq_402)\n",
    "                flank3_404 = line.find(flank3_seq_404)\n",
    "                if flank5 != -1 and flank3_401 !=-1:\n",
    "                    trimmed = line[flank5+len(flank5_seq):flank3_401]\n",
    "                    dna.append(trimmed)\n",
    "                elif flank5 != -1 and flank3_402 !=-1:\n",
    "                    trimmed = line[flank5+len(flank5_seq):flank3_402]\n",
    "                    dna.append(trimmed)\n",
    "                elif flank5 != -1 and flank3_404 !=-1:\n",
    "                    trimmed = line[flank5+len(flank5_seq):flank3_404]\n",
    "                    dna.append(trimmed)\n",
    "                elif line.find(flank_stop_splicevar) !=-1:\n",
    "                    dna.append('*splicevar')\n",
    "\n",
    "    data = pd.DataFrame({'dna':dna})\n",
    "\n",
    "    #collapse to unique sequences with counts\n",
    "    data_uniq = data['dna'].value_counts()\n",
    "    data_uniq = pd.DataFrame(data_uniq).reset_index()\n",
    "    data_uniq.columns = ['dna','count']\n",
    "    \n",
    "    print('---')\n",
    "    print('plex:',plex)\n",
    "    print(\"number of contigs:\",number_contigs)\n",
    "    print(\"number of contigs with correct flanking sequence:\", len(data))\n",
    "    print(\"number of unique contigs:\",len(data_uniq))\n",
    "\n",
    "    return data_uniq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "---\n",
      "plex: 7096\n",
      "number of contigs: 349911\n",
      "number of contigs with correct flanking sequence: 147523\n",
      "number of unique contigs: 63406\n",
      "---\n",
      "plex: 7097\n",
      "number of contigs: 373245\n",
      "number of contigs with correct flanking sequence: 165281\n",
      "number of unique contigs: 68991\n",
      "---\n",
      "plex: 7098\n",
      "number of contigs: 315287\n",
      "number of contigs with correct flanking sequence: 144975\n",
      "number of unique contigs: 66583\n",
      "---\n",
      "plex: 7099\n",
      "number of contigs: 322062\n",
      "number of contigs with correct flanking sequence: 175865\n",
      "number of unique contigs: 32078\n",
      "---\n",
      "plex: 7100\n",
      "number of contigs: 354255\n",
      "number of contigs with correct flanking sequence: 177871\n",
      "number of unique contigs: 25635\n",
      "---\n",
      "plex: 7101\n",
      "number of contigs: 342865\n",
      "number of contigs with correct flanking sequence: 186120\n",
      "number of unique contigs: 32909\n",
      "---\n",
      "plex: 7102\n",
      "number of contigs: 338488\n",
      "number of contigs with correct flanking sequence: 209115\n",
      "number of unique contigs: 20545\n",
      "---\n",
      "plex: 7103\n",
      "number of contigs: 343478\n",
      "number of contigs with correct flanking sequence: 206447\n",
      "number of unique contigs: 18083\n",
      "---\n",
      "plex: 7104\n",
      "number of contigs: 383959\n",
      "number of contigs with correct flanking sequence: 269979\n",
      "number of unique contigs: 24368\n",
      "---\n",
      "plex: 7105\n",
      "number of contigs: 398695\n",
      "number of contigs with correct flanking sequence: 276010\n",
      "number of unique contigs: 21087\n",
      "---\n",
      "plex: 7106\n",
      "number of contigs: 375170\n",
      "number of contigs with correct flanking sequence: 259776\n",
      "number of unique contigs: 19537\n",
      "---\n",
      "plex: 7107\n",
      "number of contigs: 380324\n",
      "number of contigs with correct flanking sequence: 256567\n",
      "number of unique contigs: 22747\n",
      "---\n",
      "plex: 7108\n",
      "number of contigs: 250743\n",
      "number of contigs with correct flanking sequence: 185557\n",
      "number of unique contigs: 15798\n",
      "---\n",
      "plex: 7109\n",
      "number of contigs: 341961\n",
      "number of contigs with correct flanking sequence: 270286\n",
      "number of unique contigs: 19855\n",
      "---\n",
      "plex: 7110\n",
      "number of contigs: 329349\n",
      "number of contigs with correct flanking sequence: 261113\n",
      "number of unique contigs: 21016\n",
      "---\n",
      "plex: 7111\n",
      "number of contigs: 267762\n",
      "number of contigs with correct flanking sequence: 196769\n",
      "number of unique contigs: 28862\n",
      "---\n",
      "plex: 7112\n",
      "number of contigs: 378058\n",
      "number of contigs with correct flanking sequence: 234866\n",
      "number of unique contigs: 33444\n",
      "---\n",
      "plex: 7113\n",
      "number of contigs: 349661\n",
      "number of contigs with correct flanking sequence: 189106\n",
      "number of unique contigs: 29426\n",
      "---\n",
      "plex: 7114\n",
      "number of contigs: 408317\n",
      "number of contigs with correct flanking sequence: 296605\n",
      "number of unique contigs: 30098\n",
      "---\n",
      "plex: 7115\n",
      "number of contigs: 478667\n",
      "number of contigs with correct flanking sequence: 334812\n",
      "number of unique contigs: 34333\n",
      "---\n",
      "plex: 7116\n",
      "number of contigs: 402638\n",
      "number of contigs with correct flanking sequence: 273113\n",
      "number of unique contigs: 27496\n",
      "---\n",
      "plex: 7117\n",
      "number of contigs: 293094\n",
      "number of contigs with correct flanking sequence: 207791\n",
      "number of unique contigs: 20737\n",
      "---\n",
      "plex: 7118\n",
      "number of contigs: 287788\n",
      "number of contigs with correct flanking sequence: 182749\n",
      "number of unique contigs: 29048\n",
      "---\n",
      "plex: 7119\n",
      "number of contigs: 251017\n",
      "number of contigs with correct flanking sequence: 7358\n",
      "number of unique contigs: 5230\n",
      "---\n",
      "plex: 7120\n",
      "number of contigs: 381846\n",
      "number of contigs with correct flanking sequence: 264567\n",
      "number of unique contigs: 27038\n",
      "---\n",
      "plex: 7121\n",
      "number of contigs: 313406\n",
      "number of contigs with correct flanking sequence: 214170\n",
      "number of unique contigs: 21487\n",
      "---\n",
      "plex: 7122\n",
      "number of contigs: 397196\n",
      "number of contigs with correct flanking sequence: 276949\n",
      "number of unique contigs: 24892\n",
      "---\n",
      "plex: 7123\n",
      "number of contigs: 277531\n",
      "number of contigs with correct flanking sequence: 192863\n",
      "number of unique contigs: 23337\n",
      "---\n",
      "plex: 7124\n",
      "number of contigs: 227750\n",
      "number of contigs with correct flanking sequence: 154359\n",
      "number of unique contigs: 21174\n",
      "---\n",
      "plex: 7125\n",
      "number of contigs: 373427\n",
      "number of contigs with correct flanking sequence: 305012\n",
      "number of unique contigs: 28573\n",
      "---\n",
      "plex: 7126\n",
      "number of contigs: 282530\n",
      "number of contigs with correct flanking sequence: 214434\n",
      "number of unique contigs: 22218\n",
      "---\n",
      "plex: 7127\n",
      "number of contigs: 327031\n",
      "number of contigs with correct flanking sequence: 260855\n",
      "number of unique contigs: 23232\n",
      "---\n",
      "plex: 7128\n",
      "number of contigs: 350540\n",
      "number of contigs with correct flanking sequence: 281391\n",
      "number of unique contigs: 24217\n",
      "---\n",
      "plex: 7129\n",
      "number of contigs: 238761\n",
      "number of contigs with correct flanking sequence: 198628\n",
      "number of unique contigs: 22803\n",
      "---\n",
      "plex: 7130\n",
      "number of contigs: 198808\n",
      "number of contigs with correct flanking sequence: 152677\n",
      "number of unique contigs: 19423\n",
      "---\n",
      "plex: 7131\n",
      "number of contigs: 253705\n",
      "number of contigs with correct flanking sequence: 207283\n",
      "number of unique contigs: 22547\n"
     ]
    }
   ],
   "source": [
    "for plex in [\"7096\", \"7097\", \"7098\", \"7099\", \"7100\", \"7101\", \"7102\", \"7103\", \"7104\", \"7105\", \"7106\",\n",
    "             \"7107\", \"7108\", \"7109\", \"7110\", \"7111\", \"7112\", \"7113\", \"7114\", \"7115\", \"7116\", \"7117\",\n",
    "             \"7118\", \"7119\", \"7120\", \"7121\", \"7122\", \"7123\", \"7124\", \"7125\", \"7126\", \"7127\", \"7128\",\n",
    "             \"7129\", \"7130\", \"7131\"]:\n",
    "    data_uniq = process(plex)\n",
    "    data_uniq = data_uniq.rename(columns={\"count\":'count_'+plex_dict[plex]})\n",
    "    merged = merged.merge(data_uniq, left_on='dna', right_on='dna', how='outer')\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "#identify peptides that match stop codon-containing NNN peptides from the library for doping into\n",
    "## pattern: \"NNNTAANNNNNNNNNTAGNNNNNNNNNNNNTGANNNNNN\"\n",
    "\n",
    "def check_doped(trimmed):\n",
    "    if len(trimmed) == 39:\n",
    "        if trimmed.find('TAA') == 3:\n",
    "            if trimmed.find('TAG') == 15:\n",
    "                if trimmed.find('TGA') == 30:\n",
    "                    return 'match'\n",
    "                else:\n",
    "                    return float('NaN')\n",
    "            else:\n",
    "                return float('NaN')\n",
    "        else:\n",
    "            return float('NaN')\n",
    "    else:\n",
    "        return float('NaN')  \n",
    "    \n",
    "dna_list = merged['dna']\n",
    "n = len(dna_list)\n",
    "doped = ['']*n\n",
    "\n",
    "for i in range(0,n):\n",
    "    doped[i] = check_doped(dna_list[i])\n",
    "\n",
    "merged['doped_match'] = doped\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "#translate DNA\n",
    "aa = []\n",
    "for dna in merged['dna']:\n",
    "    if len(dna)%3 == 0:\n",
    "        aa.append(translate(dna))\n",
    "    else:\n",
    "        aa.append('***DNA Not Mult of 3')\n",
    "merged['aa'] = aa\n",
    "    \n",
    "merged.to_csv('twist_covid.csv')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "624638\n",
      "167865\n"
     ]
    }
   ],
   "source": [
    "#filter for peptides that don't match pattern of doped library or Twist order\n",
    "print(len(merged))\n",
    "merged_filtered = merged.dropna(subset=['name', 'doped_match'], how='all')\n",
    "print(len(merged_filtered))\n",
    "merged_filtered.to_csv('twist_covid_filtered.csv')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
